### ISSUE TITLE
Unable to start Glowroot on Wildfly 24.0.1

### ISSUE THREAD

**Issue by MrAligual** [glowroot/glowroot#1045]
I'm trying to install Glowroot on Wildfly 24.0.1 with java-1.8.0-openjdk. 

I've tried with the Embedded Collector and Central Collector Versions, and with 0.13.6 or newer but I get the following error no matter what I do on glowroot log:

`2023-10-20 09:58:37.423 INFO  org.glowroot - Glowroot version: 0.13.6, built 2020-03-01 01:25:31 +0000
2023-10-20 09:58:37.426 INFO  org.glowroot - Java version: 1.8.0_382 (Red Hat, Inc. / Linux)
2023-10-20 09:58:37.427 INFO  org.glowroot - Java args: -Xms4000m -Xmx4000m -XX:MetaspaceSize=512m -XX:MaxMetaspaceSize=1024m -verbose:gc -Xloggc:/opt/wildfly01/standalone/log/gc.log -XX:+PrintGCDetails -XX:+PrintGCDateStamps -XX:+UseGCLogFileRotation -XX:NumberOfGCLogFiles=5 -XX:GCLogFileSize=10M -javaagent:/opt/glowroot/glowroot.jar -agentlib:jdwp=transport=dt_socket,address=8787,server=y,suspend=n
2023-10-20 09:58:42.696 INFO  org.glowroot - creating glowroot schema...
2023-10-20 09:58:42.965 INFO  org.glowroot - glowroot schema created
2023-10-20 09:58:43.481 INFO  org.glowroot - UI listening on 127.0.0.1:4000 (to access the UI from remote machines, change the bind address to 0.0.0.0, either in the Glowroot UI under Configuration > Web or directly in the admin.json file, and then restart JVM to take effect)
2023-10-20 09:59:16.397 ERROR o.g.a.p.servlet.ContainerStartup_ - could not create platform mbean server: java/lang/management/ManagementFactory
java.lang.NoClassDefFoundError: java/lang/management/ManagementFactory
        at org.glowroot.agent.plugin.servlet.ContainerStartup_.initPlatformMBeanServer(ContainerStartup.java:58)
        at org.glowroot.agent.plugin.servlet.ServletAspect$ServiceInitAdvice_.onBefore(ServletAspect.java:439)
        at javax.servlet.GenericServlet.init(GenericServlet.java)
        at io.undertow.servlet.core.LifecyleInterceptorInvocation.proceed(LifecyleInterceptorInvocation.java:117)
        at org.wildfly.extension.undertow.security.RunAsLifecycleInterceptor.init(RunAsLifecycleInterceptor.java:78)
        at io.undertow.servlet.core.LifecyleInterceptorInvocation.proceed(LifecyleInterceptorInvocation.java:103)
        at io.undertow.servlet.core.ManagedServlet$DefaultInstanceStrategy.start(ManagedServlet.java:305)
        at io.undertow.servlet.core.ManagedServlet.createServlet(ManagedServlet.java:145)
        at io.undertow.servlet.core.DeploymentManagerImpl$2.call(DeploymentManagerImpl.java:588)
        at io.undertow.servlet.core.DeploymentManagerImpl$2.call(DeploymentManagerImpl.java:559)
        at io.undertow.servlet.core.ServletRequestContextThreadSetupAction$1.call(ServletRequestContextThreadSetupAction.java:42)
        at io.undertow.servlet.core.ContextClassLoaderSetupAction$1.call(ContextClassLoaderSetupAction.java:43)
        at org.wildfly.extension.undertow.security.SecurityContextThreadSetupAction.lambda$create$0(SecurityContextThreadSetupAction.java:105)
        at org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1535)
        at org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1535)
        at org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1535)
        at org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1535)
        at io.undertow.servlet.core.DeploymentManagerImpl.start(DeploymentManagerImpl.java:601)
        at org.wildfly.extension.undertow.deployment.UndertowDeploymentService.startContext(UndertowDeploymentService.java:106)
        at org.wildfly.extension.undertow.deployment.UndertowDeploymentService$1.run(UndertowDeploymentService.java:87)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
        at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:1990)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1486)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1377)
        at java.lang.Thread.run(Thread.java:750)
        at org.jboss.threads.JBossThread.run(JBossThread.java:513)
Caused by: java.lang.ClassNotFoundException: java.lang.management.ManagementFactory from [Module "javax.servlet.api" version 2.0.0.Final from local module loader @6b00f608 (finder: local module finder @2f17e30d (roots: /opt/wildfly01/modules,/opt/wildfly01/modules/system/layers/base))]
        at org.jboss.modules.ModuleClassLoader.findClass(ModuleClassLoader.java:255)
        at org.jboss.modules.ConcurrentClassLoader.performLoadClassUnchecked(ConcurrentClassLoader.java:410)
        at org.jboss.modules.ConcurrentClassLoader.performLoadClass(ConcurrentClassLoader.java:398)
        at org.jboss.modules.ConcurrentClassLoader.loadClass(ConcurrentClassLoader.java:116)
        ... 28 common frames omitted
`

**Comment by MrAligual** [comment]
My mistake, glowroot is indeed running but I get that message on the log. Any clues why?

**Comment by majlantalik** [comment]
I get the same error, even though Glowroot itself works. Unfortunately, this make it unusable for production use for me. Any clue as to how to solve this?

I'm running Glowroot on Wildfly 26.1.2.Final and it behaves the same with any jdk-11 distribution I've tried.

**Comment by smalltox** [comment]
see https://github.com/glowroot/glowroot/issues/1025

**Comment by Never-lab** [comment]
Hi @MrAligual — thanks for the follow-up that Glowroot does start; that matches what we see in the code.

## What the ERROR actually means

It is **not** a failed agent bootstrap. During servlet init the servlet plugin calls `ManagementFactory.getPlatformMBeanServer()` so gauges / `LazyPlatformMBeanServer` can attach once WildFly is ready.

On JBoss Modules that call often runs under the **`javax.servlet.api`** (or jakarta servlet API) module classloader. That module typically has **no dependency on `java.management`**, so you get:

`ClassNotFoundException: java.lang.management.ManagementFactory from [Module "javax.servlet.api" …]`

Glowroot catches the throwable and logs `ERROR … could not create platform mbean server` — the JVM and agent keep running. Annoying for production logs, but usually not “Glowroot won’t start.”

## Practical workaround (WildFly / EAP)

Add a dependency on `java.management` to the servlet API module (community reports on #1025, including newer WildFly):

Edit something like:

- `$WILDFLY/modules/system/layers/base/javax/servlet/api/main/module.xml`  
  or, on jakarta-based installs:  
  `…/jakarta/servlet/api/main/module.xml`

and ensure:

```xml
<dependencies>
  <module name="java.management"/>
</dependencies>
```

Then restart WildFly. (If `java.management` is missing as a module on very custom installs, people have had to add/reference a small module that depends on `javax.api` / the JDK management APIs — details vary by WildFly version.)

Also unrelated but useful on EAP/WildFly if you later hit weaving/`GeneratedMethodMeta` issues: `-Djboss.modules.system.pkgs=org.glowroot` (see #1112).

## Related reports

Same stack / same topic: **#1025** (most discussion + workarounds), #956, #596, #1098. Worth treating those as one cluster.

If the ERROR remains after the `java.management` module dependency, please say which exact `module.xml` you edited and WildFly/EAP + Glowroot versions — happy to refine the hint.

cc @nowheresly @trask

**Comment by Never-lab** [comment]
Already covered here and in **#1025**: `ContainerStartup` → `ManagementFactory` under WildFly’s `javax.servlet.api` module → noisy ERROR; Glowroot still runs (as you confirmed). Workaround: add `java.management` dependency for the servlet API module (see #1025). Not “unable to start” once that misunderstanding is cleared.

@nowheresly — please close as duplicate of #1025 / support.
